util/store/store.go

```diff
@@ -0,0 +1,194 @@
+package store
+
+import (
+	"fmt"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	gogotypes "github.com/gogo/protobuf/types"
+)
+
+// GetInt retrieves an sdkmath.Int from a KVStore, or a provided minimum value if no value is stored.
+// It panics if a stored value fails to unmarshal or is less than or equal to the minimum value.
+// Accepts an additional string which should describe the field being retrieved in custom error messages.
+func GetInt(store sdk.KVStore, key []byte, min sdkmath.Int, desc string) sdkmath.Int {
+	if bz := store.Get(key); bz != nil {
+		val := sdk.ZeroInt()
+		if err := val.Unmarshal(bz); err != nil {
+			panic(fmt.Sprintf("error unmarshaling %s into %T: %s", desc, val, err))
+		}
+		if val.IsNil() || val.LTE(min) {
+			panic(fmt.Sprintf("%s is not above the minimum %s of %s", val, desc, min))
+		}
+		return val
+	}
+	// No stored bytes at key: return minimum value
+	return min
+}
+
+// SetInt stores an sdkmath.Int in a KVStore, or clears if setting to a provided minimum value or nil.
+// Returns an error on attempting to store value lower than the minimum or on failure to encode.
+// Accepts an additional string which should describe the field being retrieved in custom error messages.
+func SetInt(store sdk.KVStore, key []byte, val, min sdkmath.Int, desc string) error {
+	if val.IsNil() || val.Equal(min) {
+		store.Delete(key)
+		return nil
+	}
+	if val.LT(min) {
+		return fmt.Errorf("%s is below the minimum %s of %s", val, desc, min)
+	}
+	bz, err := val.Marshal()
+	if err != nil {
+		return err
+	}
+	store.Set(key, bz)
+	return nil
+}
+
+// GetDec retrieves an sdk.Dec from a KVStore, or a provided minimum value if no value is stored.
+// It panics if a stored value fails to unmarshal or is less than or equal to the minimum value.
+// Accepts an additional string which should describe the field being retrieved in custom error messages.
+func GetDec(store sdk.KVStore, key []byte, min sdk.Dec, desc string) sdk.Dec {
+	if bz := store.Get(key); bz != nil {
+		val := sdk.ZeroDec()
+		if err := val.Unmarshal(bz); err != nil {
+			panic(fmt.Sprintf("error unmarshaling %s into %T: %s", desc, val, err))
+		}
+		if val.IsNil() || val.LTE(min) {
+			panic(fmt.Sprintf("%s is not above the minimum %s of %s", val, desc, min))
+		}
+		return val
+	}
+	// No stored bytes at key: return minimum value
+	return min
+}
```

Comment:
turned this into a github issue since they resisted my initial attempts

(unmarshal uses a pointer receiver, the zero values `sdk.ZeroInt` and `sdk.ZeroDec` are not the same as initializing (nil) those types, and several other issues made it not worth the lines it would save)